packages/admin/admin/src/form/useAsyncAutocompleteOptionsProps.ts

```diff
@@ -0,0 +1,77 @@
+import { type UseAutocompleteProps } from "@mui/material";
+import { debounce } from "@mui/material/utils";
+import { type ChangeEvent, useCallback, useMemo, useState } from "react";
+
+export interface AsyncAutocompleteOptionsProps<T> {
+    isAsync: boolean;
+    open: boolean;
+    options: T[];
+    loadingError: Error | null;
+    loading?: boolean;
+    onOpen: (event: ChangeEvent) => void;
+    onClose: (event: ChangeEvent) => void;
+    onInputChange: UseAutocompleteProps<any, any, any, any>["onInputChange"];
+}
+
+/**
+ * A hook that handles the state management and loading behavior for asynchronous autocomplete fields.
+ * It returns a set of props ready to be used with FinalFormAsyncAutocomplete component.
+ */
+export function useAsyncAutocompleteOptionsProps<T>(loadOptions: (search?: string) => Promise<T[]>): AsyncAutocompleteOptionsProps<T> {
+    const [open, setOpen] = useState(false);
+    const [options, setOptions] = useState<T[]>([]);
+    const [loading, setLoading] = useState(false);
+    const [error, setError] = useState<Error | null>(null);
+
+    const loadOptionsWithSearch = useCallback(
+        async (search?: string) => {
+            setLoading(true);
+            setError(null);
+            try {
+                const newOptions = await loadOptions(search);
+                setOptions(newOptions);
+            } catch (e) {
+                setError(e);
+            } finally {
+                setLoading(false);
+            }
+        },
+        [loadOptions],
+    );
+
+    const debouncedLoadOptionWithSearch = useMemo(
+        () =>
+            debounce(async (value: string) => {
+                await loadOptionsWithSearch(value);
+            }, 500),
+        [loadOptionsWithSearch],
+    );
```

Comment:
For other use cases we use useDebouncedCallback, should we use that here as well? https://usehooks-ts.com/react-hook/use-debounce-callback